Hybrid key management and KEM algorithm support

paulidale opened the pull request:
Hybrid KEM and key management algorithm.

- [x] documentation is added or updated
- [x] tests are added or updated

t-j-h commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,301 @@
+    return ossl_hybrid_kem_init(ctx, params, &EVP_PKEY_decapsulate_init);
+}
+
+int ossl_hybrid_kem_encaps_init(HYBRID_PKEY_CTX *ctx, HYBRID_PKEY *authpriv,
```
calling this encaps_init when you pass all the inits through it (encaps and decaps and the auth variants) seems like a poor naming choice ...

FdaSilvaYY commented:
Typo in one commit message : Add param helpers to dupicate sans items and multi-merges

petrovr commented:
Hello,

На 19.11.24 г. в 5:55 ч., Pauli написа:
> Hybrid KEM and key management algorithm.

[SNIP]

Currently I test hybrid KEM based on existing EVP PKEY functionality and 
I cannot understand why is needed new API.


Regards,
Roumen Petrov

paulidale commented:
This is for integration into TLS where only one algorithm can be specified for KEM.

slontis commented on providers/common/include/prov/hybrid.h:
```diff
@@ -0,0 +1,52 @@
+typedef struct hybrid_st HYBRID_PKEY;
+typedef struct hybrid_ctx_st HYBRID_PKEY_CTX;
+
+struct hybrid_st {
```
hybrid_pkey_st & hybrid_pkey_ctx_st

slontis commented on providers/implementations/kem/25519_448_kem.c:
```diff
@@ -0,0 +1,13 @@
+/*
```
Rename to X.... same in keymanager

paulidale replied:
??? There is no 'x' in the key manager equivalent.
This is also transient and won't make the final cut.

slontis commented on providers/common/include/prov/hybrid.h:
```diff
@@ -0,0 +1,52 @@
+/*
```
rename to hybrid_pkey

slontis commented on providers/implementations/kem/hybrid_kem.inc:
```diff
@@ -0,0 +1,62 @@
+#include "prov/provider_ctx.h"
+#include "prov/providercommon.h"
+
+#if !defined(NAME_1) || !defined(NAME_2)
```
hybrid_defines already does this line...

slontis commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,300 @@
+    ctx->names = names;
+    ctx->key_lengths = key_lengths;
+    for (i = 0; i < num_algs; i++)
+        if (0)
```
?

slontis commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,300 @@
+        return 0;
+    /* Calculate lengths */
+    for (i = 0; i < ctx->num_algs; i++) {
+        if (!EVP_PKEY_encapsulate(ctx->ctxs[i], NULL, i + elen,
```
elen and slen are strange names.. Even stranger is the order you do i + elen which makes it not look like an array offset

paulidale replied:
enclen and secretlen were taken...

slontis commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,300 @@
+        *secretlen = secretsum;
+        return 1;
+    }
+    if (*secretlen < secretsum) {
```
Assumption is secretlen cant be NULL.

paulidale replied:
Yes.  You won't know the output length otherwise.

paulidale replied:
Is this a problem?

The size needs to be passed in and returned?  It doesn't make sense for it to be NULL.

baentsch replied:
This is the [normal pattern for all these APIs](https://docs.openssl.org/master/man3/EVP_PKEY_decapsulate/#examples), no? Pass in NULL data (`secret` in this case) and use `secretlen` to find out actual length. Rince-repeat (2 times in this case). fwiw, lgtm.

slontis commented on providers/defltprov.c:
```diff
@@ -476,6 +476,10 @@ static const OSSL_ALGORITHM deflt_asym_kem[] = {
     { PROV_NAMES_EC, "provider=default", ossl_ec_asym_kem_functions },
+# ifndef OPENSSL_NO_ECX
+    /* Temporary until PQC hybrids exist */
+    { "X25519-X448", "provider=default", ossl_X25519_X448_kem_functions },
```
Should hybrid be in the name?

paulidale replied:
If it were going to stay in.  It's never getting merged.

paulidale replied:
For a transient algorithm that is not going to be merged, I don't see the need :grin:

baentsch replied:
Please note that naming according to https://www.ietf.org/archive/id/draft-kwiatkowski-tls-ecdhe-mlkem-02.html is defined to be X25519MLKEM768 and SecP256r1MLKEM768. If I understand the hybrid algorithm instantiation logic right, it depends on a lexical separation of name components which seems to be not guaranteed. Also, please note that ML-KEM-768 already is in the feature branch by now and could also be used for testing instead of this originally proposed artificial hybrid.

baentsch replied:
Ahh, scrap that comment: Just saw that this code also depends on some constants not yet in the feature branch's includes.

paulidale replied:
Yeah, I went to add the real hybrid algs earlier today and discovered that they'd not made it in yet :)

The naming isn't restricted here.  The sub-algorithms must be listed in the order they appear in the concatenated output but this bears no relation to the overall algorithm name ordering etc.

slontis commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,667 @@
+        for (i = 0; i < ctx->num_algs; i++)
+            if ((priv && !EVP_PKEY_private_check(ctx->ctxs[i]))
+                    || (pub && !EVP_PKEY_public_check(ctx->ctxs[i])))
+                return 0;
```
pairwise test also if they are both set?

slontis replied:
Not sure if fast/slow will apply to these?

paulidale replied:
Good idea.

slontis commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,667 @@
+                return 0;
+            break;
+
+        case PIDX_PKEY_PARAM_ENCODED_PUBLIC_KEY:
```
This assumes the algs support encoded_public_key..

paulidale replied:
It does.  Both ECX algs do.

I'm not sure how to handle such error cases nicely.

vdukhovni replied:
FWIW, ML-KEM supports this.

slontis commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,667 @@
+        case PIDX_PKEY_PARAM_ENCODED_PUBLIC_KEY:
+            t = ossl_hybrid_get_key_length_from_ctx(ctx);
+            if (t > p->data_size) {
+                p->data_size = t;
```
why set to t here on error?

paulidale replied:
Returning the required size is how params are meant to work.  I.e. you got an error and this is the required size.

mattcaswell replied:
But surely you must do this only after the error below is raised? Otherwise the sizes in the error message are wrong.

slontis commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,667 @@
+        type = ossl_param_find_pidx(p->key);
+        switch (type) {
+        case PIDX_PKEY_PARAM_SECURITY_BITS:
+            t = maximum_ctx_uint32_params(ctx->num_algs, ctx->ctxs,
```
Not sure if its the max or the min of the algorithms.

paulidale replied:
It's the max.  If one alg is secure, the hybrid is.

slontis commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,667 @@
+    if (!ctx_subalg_set_param(ctx->num_algs, ctx->ctxs, params))
+        return 0;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PROPERTIES);
```
Maybe propq should be done first?

paulidale replied:
It is done first in the sub algs.  This is keeping a local copy.

slontis commented on crypto/params_dup.c:
```diff
@@ -234,3 +235,73 @@ void OSSL_PARAM_free(OSSL_PARAM *params)
+{
+    OSSL_PARAM *r, *p;
+    const OSSL_PARAM *q;
+    size_t i, m = 1;
```
Comment /* Add one for the terminator */

slontis commented on include/internal/params.h:
```diff
@@ -36,3 +40,29 @@ int ossl_param_get1_octet_string(const OSSL_PARAM *params, const char *name,
+/*
+ * Check that a param list is NULL or empty.
+ */
+static ossl_inline int ossl_param_is_empty(const OSSL_PARAM params[])
```
There is another one of these declared in the provider includes.

paulidale replied:
There isn't anymore.  I moved this to a more visible header.

slontis commented:
Looks like a reasonable approach overall.

paulidale commented on crypto/params_dup.c:
```diff
@@ -234,3 +235,73 @@ void OSSL_PARAM_free(OSSL_PARAM *params)
+    return n;
+}
+
+OSSL_PARAM *ossl_param_dup_omit(const OSSL_PARAM *src, OSSL_PARAM *omit)
```
It's from an earlier version of the code that handled params differently.  Will remove.

mattcaswell commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,293 @@
+        return 0;
+    /* Calculate lengths */
+    for (i = 0; i < ctx->num_algs; i++) {
+        if (!EVP_PKEY_encapsulate(ctx->ctxs[i], NULL, seclens + i,
```
decapsulate? Or perhaps you really meant encapsulate here in order to calculate the enclens. In which case a comment would be really nice here. It's quite surprising to see decapsulate calling encapsulate

paulidale replied:
My mistake, not that it matters.

paulidale replied:
Comment added.

mattcaswell commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,293 @@
+    if (ctx == NULL)
+        return;
+    for (i = 0; i < ctx->num_algs; i++)
+        EVP_PKEY_CTX_free(ctx->ctxs[i]);
```
I don't see anywhere where you create these `EVP_PKEY_CTX`s. What did I miss?

paulidale replied:
nothing.

mattcaswell commented on providers/common/include/prov/hybrid_pkey.h:
```diff
@@ -0,0 +1,49 @@
+typedef struct {
+    OSSL_LIB_CTX *libctx;
+    char *propq;
+    CRYPTO_REF_COUNT references;
```
We don't seem to have an up_ref anywhere. So why do we need this?

mattcaswell commented on providers/implementations/keymgmt/hybrid_kmgmt.inc:
```diff
@@ -0,0 +1,81 @@
+};
+
+/* The length of each key, the total length is in [NUM_ALGS] */
+uint32_t COMMON_NAME(key_lengths)[NUM_ALGS + 1];
```
This might cause problems for tsan. Multiple libctx's both of which have this keymanager loaded are going to fight over populating this. Presumably they are going to populate it with the same answer, but if the writes are non-atomic it could in theory cause problems.

paulidale replied:
It will always be the same value.  I'd thought about threading implications but hadn't acted yet.  Any solution is going to introduce a performance hit unfortunately.

mattcaswell replied:
Are we over-engineering this by attempting a run time calculation of the key lengths? We should know these values a compile time. Can't we just fix these values then and trust (or assert) that they are equal to those specified in the individual algorithms? We can therefore avoid any performance hit and the threading complexities.

baentsch replied:
While this doesn't quite allow for the originally intended "dynamic" nature of truly using (only) EVP APIs I'd somewhat "sympathize" with the comment by @mattcaswell : The [draft spec](https://datatracker.ietf.org/doc/draft-kwiatkowski-tls-ecdhe-mlkem/) this code imo mostly should consider also computes fixed sizes for all hybrids. So, if pulling lengths from the constituent algorithms creates more problems than it solves, I'd also be fine "streamlining" things to by-default supported KEM algs (and their known sizes, incl. ML-KEM, of course) only. Maybe in combination with a "TODO(future KEM algs)" reminder that this is a shortcut (?).

paulidale replied:
As if any of us would over engineer a solution :)

Adding the lengths to the algorithm files (`25519_448_*.c`) is easy enough and doesn't complicate things from the users' perspective.

We'll still have the threading issues with the various gettables and settables in the PKEY object.  Adding these to the algorithm files won't be so nice but I'll have a think.

mattcaswell commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,668 @@
+            s->key_lengths[MAX_HYBRID_ALGS] =                               \
+                accumulate_ ## lname ## _uint32_params                      \
+                    (s->num_algs, s->lname ## s,                            \
+                     OSSL_PKEY_PARAM_MAX_SIZE, s->key_lengths);             \
```
We're calling this "key lengths" but the param is `OSSL_PKEY_PARAM_MAX_SIZE` which seems odd

paulidale replied:
Yeah, bad nomenclature when the param name was chosen but it's too late to go back and fix it.

mattcaswell commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,668 @@
+        for (i = 0; i < key->num_algs; i++)
+            pary[i] = EVP_PKEY_gettable_params(key->keys[i]);
+        pary[i] = hybrid_gettable_params;
+        key->gettable = ossl_param_merge(key->num_algs + 1, pary);
```
`EVP_PKEY`s can can be shared between threads since they are typically immutable once created. Looks like a potential for a race

paulidale replied:
Another case of potential multiple writes of the same thing (albeit the pointers are different, what they point to will be equal).  Also thinking above these further.

mattcaswell commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,668 @@
+        for (i = 0; i < key->num_algs; i++)
+            pary[i] = EVP_PKEY_settable_params(key->keys[i]);
+        pary[i] = hybrid_settable_params;
+        key->settable = ossl_param_merge(key->num_algs + 1, pary);
```
Possible race?

paulidale replied:
Yep.

mattcaswell commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,668 @@
+            for (i = 0; i < ctx->num_algs; i++) {
+                prms->data = q;
+                prms->data_size = ctx->key_lengths[i];
+                q += ctx->key_lengths[i];
```
Should we have an assert somewhere here to make sure we haven't accidentally run off the end of `p->data`

paulidale replied:
The check on line 494 will forbid this.
```c
            t = ossl_hybrid_get_key_length_from_ctx(ctx);
            if (t > p->data_size) {
                ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH,
                               "length must be %u bytes, have %u bytes",
                               t, p->data_size);
                p->data_size = t;
                return 0;
            }
```
I guess an assert isn't overly harmful.

paulidale commented on providers/common/include/prov/hybrid_pkey.h:
```diff
@@ -0,0 +1,49 @@
+    char *propq;
+    unsigned int num_algs;
+    EVP_PKEY_CTX *ctxs[MAX_HYBRID_ALGS];
+    uint32_t *key_lengths;
```
I figured a key length over 4Gb was slightly excessive.  It's internal so easy to change.

paulidale commented:
Is there a problem with the CIs?  `evp_test -config ../../test/default-and-legacy.cnf evppkey_rsa.txt` is failing and nothing here goes anywhere near that code.

paulidale commented:
Ready for review I think.
Note that the X25519-X448 algorithm is for testing purposes only and won't be merged.

baentsch commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,260 @@
+    if (ctx->key_length == 0)
+        return 0;
+
+    if (enc == NULL) {
```
prudent to also check secret == NULL?

paulidale replied:
I don't think this would be beneficial but I've added additional NULL checks above.

baentsch commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,260 @@
+    }
+
+    /* Set outputs */
+    *enclen = ctx->ciphertext_bytes;
```
Better done later/right before return 1 (after all constituent operations succeeded)?

paulidale replied:
Setting the output lengths correctly on a failed return is something we do do elsewhere.
I'm not too fussed either way.

baentsch commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,260 @@
+                           "secretlen or enclen changed");
+            return 0;
+        }
+        secret += ctx->info->alg[i].shared_secret_bytes;
```
If I get this right, the order of algorithms in `info->alg` defines the order of concatenations (which is OK). However, does this then require a change in logic as to how the common name is created? If I don't get it wrong, it's currently the concatenation of constituent names in that order, right? However, according to https://www.ietf.org/archive/id/draft-kwiatkowski-tls-ecdhe-mlkem-02.html, section 3.1.3, "X25519MLKEM768" then has to be declared "ML-KEM-768-X25519" (in that order) for the logic here to work correctly. The order of constituent algorithms and names for "SecP256r1MLKEM768" is unconditionally correct.

paulidale replied:
Not quite.  The order of the concatenation is defined by `info->alg` but the overall name can be anything.  That's defined separately in `providers/defltprov.c`.

t8m commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,266 @@
+        *secretlen = ctx->shared_secret_bytes;
+        return 1;
+    }
+    if (secretlen == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
```
You're already checking this condition above.

paulidale replied:
Was meant to be `secret`.

t8m commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,266 @@
+    if (ctx->key_length == 0)
+        return 0;
+
+    if (secret == NULL) {
```
Please check `secretlen != NULL` above

t8m commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,782 @@
+    HYBRID_PKEY *key;
+    unsigned int i;
+
+    if (old == NULL
+            || (key = (HYBRID_PKEY *)OPENSSL_zalloc(sizeof(*key))) == NULL)
+        return NULL;
+    if (old->propq != NULL && (key->propq = strdup(old->propq)) == NULL)
+        goto err;
+    for (i = 0; i < key->info->num_algs; i++)
+        if ((key->keys[i] = EVP_PKEY_dup(old->keys[i])) == NULL)
+            goto err;
+    key->info = old->info;
+    key->key_length = old->key_length;
+    key->bits = old->bits;
+    key->security_bits = old->security_bits;
+    return key;
```
What about the refcount? The duplicated key should have a refcount of 1.

paulidale replied:
Good catch.

t8m commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,782 @@
+             * This is treated as opaque data for the sub-algorithm.
+             *
+             * The rationale behind this encoding is that we do not know
+             * what parameters (or their associated sizes) are required for
+             * each sub-algorithm.
+             */
+            if (plen[j] == 0)
+                goto err;
+            t = (*pptr[j] & 0300) >> 6;
+            data_size = *pptr[j]++ & 077;
+            if (--plen[j] < t)
+                goto err;
+            plen[j] -= t;
+            while (t-- > 0)
+                data_size = data_size << 8 | *pptr[j]++;
+
+            /* Only include this param if it's got an associated value */
+            if (data_size > 0) {
+                if (plen[j] < data_size)
+                    goto err;
+                lprm[k] = params[j];
+                lprm[k].data = pptr[j];
+                lprm[k++].data_size = data_size;
+                pptr[j] += data_size;
+                plen[j] -= data_size;
+            }
+        }
+        /* Don't bother if there are no params */
+        if (k > 0) {
+            lprm[k] = OSSL_PARAM_construct_end();
+            ctx = EVP_PKEY_CTX_new_from_name(key->libctx,
+                                             key->info->alg[i].name,
+                                             key->propq);
+            if (ctx == NULL
+                    || EVP_PKEY_fromdata_init(ctx) <= 0
+                    || EVP_PKEY_fromdata(ctx, key->keys + i, selection,
+                                         lprm) <= 0)
+                goto err;
+        }
+    }
```
I wonder if this is the right way to do it, i.e. to join the secret and public key values of these keys into one param.

Should we instead just suffix (or prefix) the param names and merge the arrays? I.e., have `alg1_priv` and `alg2_priv` params in the resulting OSSL_PARAM array?

paulidale replied:
I started doing that and it resulted in messier code.  This is clean and simple in comparison.

t8m replied:
I disagree. Maybe my proposal would require more code but I do not think the performance is a concern here as a few strncmps should be negligible in comparison to the computations of the encaps/decaps operations.

paulidale replied:
Having tried both approaches, I disagree.

You are correct, performance isn't critical but it never hurts.

t8m commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,782 @@
+    return ret;
+}
+
+const OSSL_PARAM *ossl_hybrid_ettable_common(const OSSL_PARAM *r) {
```
Awkward name: `ossl_hybrid_getsettable_common()` would be perhaps more understandable?

paulidale replied:
Yuk :grinning:  Every name I could think of was awkard.  At least `ettable` is pronounceable.

paulidale commented:
@openssl/committers reviews please.

beldmit commented on providers/common/include/prov/hybrid_pkey.h:
```diff
@@ -0,0 +1,84 @@
+
+#define HYBRID_COMMON_FIELDS            \
+    OSSL_LIB_CTX *libctx;               \
+    char *propq;                        \
```
Could you please clarify why do we need this field and why do we have it one but not one per key?

baentsch replied:
fwiw and imo it makes sense to only have one as the "amalgamation" result representing the "joint" algorithm.

beldmit replied:
Corresponding to what? The key algos may come from different providers and have different propq...

baentsch replied:
Corresponding to the constituent algorithms' parameters -- that must be identical irrespective of which provider delivers the implementation.

paulidale replied:
We need to fetch all of the subalgs so we need a propq to use while doing so.  This is also passed through to the subalgs in case they fetch anything.  For example `ossl_hybrid_kmgmt_set_params` does this:

```c
        case PIDX_PKEY_PARAM_PROPERTIES:
            if (!OSSL_PARAM_get_utf8_string(p, &str, SIZE_MAX))
                return 0;
            OPENSSL_free(key->propq);
            key->propq = str;
            /* Fall through */

        default:
            if (!key_subalg_set_param(key->info->num_algs, key->keys, prms))
                return 0;
            break;
```

I hadn't considered a different query for the subalgs.  It is an interesting idea.  Can you think of a use case?  Would we want a different propq for each subalg or just a single one for all subalgs (that is different to the hybrid's propq).  A _nice_ implementation might be tricky but I'm sure we'd be able to brainstorm something.

baentsch replied:
How are these propq's communicated into the keymgmt logic? Doesn't that happen via "OSSL_PKEY_PARAM_PROPERTIES"? Are you suggesting that this logic should (be changed to) be able to parse a (single param) string into two (separate per-algorithm propqs)? If so, which characters are safe to use as separators for the constituent propq's? Also seems to demand some more documentation.

beldmit replied:
> Regardless, not for this PR IMO.

I agree.

BTW, how do you use propq?

paulidale replied:
One complication: where should the hybrid algorithm be located?  If it's in the FIPS provider, it then isn't possible to load anything from the P11 provider (because the FIPS provider loads only from it's internal library context that's unavailable from the outside).  This implies that it would have to be from the default provider…

The propq is used to fetch the sub-algorithms (`EVP_PKEY_CTX_new_from_name` and `EVP_PKEY_CTX_new_from_pkey`).  It is also passed through to the sub-algorithms in case they have to fetch anything.

beldmit replied:
I'm afraid it should be a property of a sub-algorithm then. But we can discuss it separately

beldmit commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,739 @@
+            break;
+
+        case PIDX_PKEY_PARAM_BITS:
+            if (!!OSSL_PARAM_set_size_t(p, key->bits))
```
`!!` ??

paulidale replied:
Typo, will fix.  Guess this one isn't covered well by our tests :disappointed:

beldmit reviewed (commented):
I don't see obvious problems, code looks reasonable. The only remaining issue is a single propq

baentsch commented:
> I don't see obvious problems, code looks reasonable. The only remaining issue is a single propq

As per comment above, some documentation may be needed, too.

My next "review" step would be to use this branch to instantiate and test the ML-KEM-768 hybrids (also to get rid of the artificial X448-X25519 test): OK for you @paulidale or premature?

paulidale commented:
My expectation is that documentation will done as part of the implementation of a real hybrid algorithm.  The reason is because the chosen hybrid algorithm(s) will need to be documented and that documentation should cover everything users require.

I could potentially add more internal documentation if this is deemed desirable.  I'm not convinced this is necessary because using this is (deliberately) fairly straightforward.

I'd like this to be merged sooner if possible -- leaving complete but unmerged stuff generally complicates things unnecessarily.  Ideally, it would be merged without the x25519-x448 hybrid (which corresponds to the commits flagged as temporary).  Nonetheless, I'd be okay merging with the x25519-x448 hybrid and removing it later in a separate PR.

baentsch commented:
> My expectation is that documentation will done as part of the implementation of a real hybrid algorithm.

OK for me unless there's "special" handling of things like special use-case support for propq required already now (is it @beldmit ?). Declaring that out of scope for me for now absolutely is reasonable.

> Nonetheless, I'd be okay merging with the x25519-x448 hybrid and removing it later in a separate PR.

Also makes sense. I'd also like to see this merged sooner than later to move forward with integration, followed by documentation and "cleanup" of this sort.

baentsch reviewed (approved):
LGTM (as per discussion on follow-up activities). Thanks @paulidale !

beldmit reviewed (approved):
LGTM

baentsch commented:
Two questions, @paulidale that emerged when trying to instantiate X25519MLKEM768 as per spec:
1) Is it possible that the "name-merging" macros don't do well with algorithm NAMEs of the form "ML-KEM-768" (the dashes, most notably)? When using "MLKEM768" everything builds great, but that's not the "desired" (standarized) name of ML-KEM-768.
2) Is the term "key_length(_bytes)" in HYBRID_ALG_INFO a bit of a mis-nomer and should rather be "encoded_pubkey_length"? Or is this indeed something that needs improving for algorithms with different pubK, privK sizes?

For reference, using the current feature-branch code base, this is the "instantiation code" I tried:

```
#define NAME_1          ML-KEM-768
#define NAME_2          X25519

#include <openssl/core_dispatch.h>
#include <openssl/core_names.h>
#include <openssl/params.h>
#include <openssl/sha.h>

#include "crypto/ecx.h"
#include "crypto/mlkem.h"
#include "prov/securitycheck.h"

#include "hybrid_kmgmt.inc"

const HYBRID_ALG_INFO COMMON_NAME(info) = {
    NUM_ALGS,
    {
        {
            "MLKEM768",
            OSSL_MLKEM768_PUBLIC_KEY_BYTES,                /* key length (bytes) */
            768,                  /* bits */
            OSSL_MLKEM768_SECURITY_BITS,         /* security bits */
            OSSL_MLKEM768_SHARED_SECRET_BYTES,       /* shared secret (bytes) */
            OSSL_MLKEM768_CIPHERTEXT_BYTES                 /* ciphertext bytes */
        },
        {
            "X25519",
            X25519_KEYLEN,              /* key length (bytes) */
            X25519_BITS,                /* bits */
            X25519_SECURITY_BITS,       /* security bits */
            SHA256_DIGEST_LENGTH,       /* shared secret (bytes) */
            X25519_KEYLEN               /* ciphertext bytes */
        },
    }
};

```

vdukhovni commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,740 @@
+    }
+}
+
+void *ossl_hybrid_kmgmt_load(const void *reference, size_t reference_sz)
```
What does this do when one of the keys is not a legacy key?  My impression from Matt is that load by reference is not a feature of solely-provided keys...

paulidale replied:
I'd guess that it doesn't get called.  I've tried dropping it and everything seems happy.  Fixup coming.

paulidale commented on providers/implementations/kem/25519_448_kem.c:
```diff
@@ -0,0 +1,29 @@
+
+const OSSL_PARAM OSSL_NAME(ctx_settable_params)[] = {
+    OSSL_PARAM_utf8_string(OSSL_KEM_PARAM_OPERATION, NULL, 0),
+    OSSL_PARAM_octet_string(OSSL_KEM_PARAM_IKME, NULL, 0),
```
At the moment the entropy is shared by all keys.  Splitting it would be straightforward.

vdukhovni replied:
Perhaps a TODO, or similar breadcrumb may be sufficient for now. At some point we may need separate `IKME`s, and then the current ones will need to be stuttered to reproduce the old tests.  The ML-KEM code (current state) expects exactly 32-bytes of IKME (and rejects over-provisioning as a parameter error), if other algorithms expect a different length, a conflict is possible.

paulidale replied:
Okay, comment added to `hybrid_kmgmt.c` rather than here (since this is going to disappear).

paulidale commented on providers/implementations/kem/ml_kem.c:
```diff
@@ -126,8 +126,8 @@ static int mlkem_set_ctx_params(void *vctx, const OSSL_PARAM params[])
-    if ((p = OSSL_PARAM_locate_const(params, OSSL_KEM_PARAM_MLKEM_ENC_ENTROPY)) != NULL
-        && (p->data_size != MLKEM_ENCAP_ENTROPY
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_KEM_PARAM_IKME)) != NULL
+        && (p->data_size < MLKEM_ENCAP_ENTROPY
```
Dropping the relevant commit causes a conflict :(

Sigh.

slontis commented on crypto/params_dup.c:
```diff
@@ -234,3 +235,12 @@ void OSSL_PARAM_free(OSSL_PARAM *params)
+{
+    size_t n;
+
+    for (n = 0; !ossl_param_is_empty(plist); plist++)
```
This is probably just better using plist->key == NULL

paulidale replied:
It an inline function, so shouldn't cost anything.  Doing it this way also covers the `plist == NULL` case without extra lines of code.

slontis commented on include/internal/params.h:
```diff
@@ -36,3 +40,17 @@ int ossl_param_get1_octet_string(const OSSL_PARAM *params, const char *name,
                                         size_t maxsize);
+
+/*
+ * Count the number of elements in a parameter list.
```
single line comments please..

paulidale replied:
Multi-line comments is the standard in this file and I've been required to change single line comments in a file that favoured multi-line comments before.  I don't mind changing this so long as nobody asks for it to be changed back…

slontis replied:
You re not following the coding standard

slontis commented on providers/common/include/prov/hybrid_pkey.h:
```diff
@@ -0,0 +1,86 @@
+    const unsigned int num_algs;
+    const struct {
+        const char *name;
+        size_t pubkey_length_bytes;
```
This looks like it is geared towards KEM.. Do you imagine this being used for signatures? Is that a separate structure for that case?
Should this be called HybridKEM?

paulidale replied:
If/when other hybrid operations are defined, I anticipate that this `info` structure will be extended to suite.

It's storing constant information about the key.  This should include everything that is relevant for any operation the key supports.  However, until that information is actually required, I see no reason including it.

Using separate structure would require separate key management since this is primarily used there.

slontis commented on providers/common/include/prov/hybrid_pkey.h:
```diff
@@ -0,0 +1,86 @@
+} HYBRID_ALG_INFO;
+
+#define HYBRID_COMMON_FIELDS            \
+    OSSL_LIB_CTX *libctx;               \
```
Is there a reason why HYBRID_COMMON_FIELDS could not be an array [2] also? (I assume its the higher level API)

paulidale replied:
It is fundamentally not an array.  It contains information about the hybrid algorithm not about the sub algorithms.  The sub algorithms' information is in the constant `info` structure which is an array.

slontis commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,276 @@
+
+static int hybrid_kem_init(HYBRID_PKEY_CTX *ctx, HYBRID_PKEY *auth,
+                           const OSSL_PARAM params[],
+                           int (*initf)(EVP_PKEY_CTX *ctx,
```
typedef this

paulidale replied:
Why???  The only functions passed in are existing and are drawn from a set of exactly two.
To typedef them properly, the typedef would need to be made public and the two functions declared using it.

It doesn't make a lot of sense IMO.

slontis commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,276 @@
+    return ossl_hybrid_set_ctx_params(ctx, params);
+}
+
+int ossl_hybrid_kem_encapsulate_init(void *vctx, void *vauth,
```
vauth is not the right name here.. There is no authentication for this case.

paulidale replied:
Copy pasted the name from elsewhere …

slontis commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,276 @@
+    return ossl_hybrid_set_ctx_params(ctx, params);
+}
+
+static int hybrid_kem_init(HYBRID_PKEY_CTX *ctx, HYBRID_PKEY *auth,
```
I dotn think this is an auth key ??

slontis commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,276 @@
+                                unsigned char *secret, size_t *secretlen)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t a, b;
```
pretty terrible names..

slontis commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,276 @@
+                           "secretlen or enclen changed");
+            return 0;
+        }
+        secret += ctx->info->alg[i].shared_secret_bytes;
```
If your name didnt suck you could do secret += b;

slontis commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,276 @@
+                                const unsigned char *enc, size_t enclen)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t a;
```
bad name

slontis commented on providers/implementations/keymgmt/build.info:
```diff
@@ -44,6 +45,12 @@ SOURCE[$KDF_GOAL]=kdf_legacy_kmgmt.c
 
+IF[{- !$disabled{ec} -}]
+  IF[{- !$disabled{ecx} -}]
+    SOURCE[$HYBRID_GOAL]=hybrid_kmgmt.c 25519_448_kmgmt.c
```
We dont usually put numbers at start of a name.  Probably would have used X25519_X448..
There would be  different name for a signature anyway

paulidale replied:
This file isn't staying, so no point "fixing" it.

slontis commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,734 @@
+        EVP_PKEY_CTX_free(ctx->ctxs[i]);
+        ctx->ctxs[i] = EVP_PKEY_CTX_new_from_name(libctx, info->alg[i].name,
+                                                  ctx->propq);
+        if (ctx->ctxs[i] == NULL)
```
if i is 1 should this goto err and clean itself up here?

paulidale replied:
I guess so.

slontis commented on crypto/params_dup.c:
```diff
@@ -234,3 +235,12 @@ void OSSL_PARAM_free(OSSL_PARAM *params)
+
+    for (n = 0; !ossl_param_is_empty(plist); plist++)
+        n++;
+    return n;
```
Should this be +1 for the terminator.. (I guess you are treating this like strlen)

paulidale replied:
Yep.

slontis commented on test/recipes/30-test_evp_data/evppkey_ecx_kem.txt:
```diff
@@ -0,0 +1,55 @@
+# Decapsulation KATs for ML-KEM-768 from BoringSSL
+# https://boringssl.googlesource.com/boringssl/+/refs/heads/master/crypto/mlkem/mlkem768_decap_tests.txt
+
+# Keys genreated using:
```
generated

slontis commented on test/recipes/30-test_evp_data/evppkey_ecx_kem.txt:
```diff
@@ -0,0 +1,55 @@
+Input = 37b7d2a26277d6d00dfcf8c1eaa8d72711e1a9c1a4776625b00d51aaf17b773bb900d9abd1f6385997313f9b6ed9ad85e734134b1e4e7ec3
+Output = aa6dd78e6be63b438518da22351e2c60e2caca0725551b0d17a1b0ce8dff6199e9e793518d1264b4074fc0cd9d867bf611456d094c8b4b3f60b3837d174cb8ff
+
+Title = Hybrid ECX/ECX KATs
```
Would probably put these is a separate file.

paulidale replied:
It's temporary and being removed.  It's fine here.

slontis commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,734 @@
+    return 1;
+}
+
+int ossl_hybrid_kmgmt_export(void *keydata, int selection,
```
I find your lack of comments here disturbing :)

slontis commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,734 @@
+int ossl_hybrid_kmgmt_import(void *keydata, int selection,
+                             const OSSL_PARAM params[])
+{
+    HYBRID_PKEY *key = (HYBRID_PKEY *)keydata;
```
This format seems to imply that exported data is not persisted.. It seems any key string added would break the import?

paulidale replied:
Yep.  This is a purely transient format so that the OSSL_PARAMs from different algorithms can be handled sanely.

At some point in the future, we might end up with encoders and decoders which would deal with non-transient keying.

slontis reviewed (commented):
Looks good overall.. I have completed reviewing.

paulidale commented:
Feedback addressed.

paulidale commented:
@beldmit, cosmetic changes since your approval but they are large enough in scope to merit a reconfirmation.

beldmit commented on providers/implementations/kem/hybrid_kem.c:
```diff
@@ -0,0 +1,277 @@
+        if (ciphertext_bytes != ctx->info->alg[i].ciphertext_bytes
+                || shared_secret_bytes != ctx->info->alg[i].shared_secret_bytes) {
+            ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH,
+                           "secretlen or enclen changed");
```
Shouldn't it be 2 separate diagnostics?

paulidale replied:
Given that this failure should never happen, I'm not sure how useful separate diagnostics would be.

If either of the conditions occurs it indicates a serious internal consistency problem.  Essentially, the algorithm lied about the sizes.

I can split this into two errors or, perhaps better, change it to an internal error if you insist.

beldmit replied:
Yes, then internal error would be better

paulidale replied:
I've done this too.

beldmit commented:
I don't see obvious problems with the code.

beldmit commented:
My approval persists

t8m commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,746 @@
+            OSSL_PARAM_set_all_unmodified(prms);
+            for (i = 0; i < key->info->num_algs; i++) {
+                prms->data = q;
+                prms->data_size = key->info->alg[i].pubkey_length_bytes;
+                q += key->info->alg[i].pubkey_length_bytes;
```
priv key will have different length than pub key, won't it?

t8m commented on providers/implementations/keymgmt/hybrid_kmgmt.c:
```diff
@@ -0,0 +1,746 @@
+                             &hybrid_export_cb, &arg))
+            goto err;
+
+    /* Package each OSSL_PARAM */
```
So export/import uses different format for the priv/pub params than get_params/set_params. That is really horrible. What about not supporting priv and pub in get/set_params at all? And just support the encoded public key param there. That should be sufficient for TLS, shouldn't it?

paulidale replied:
Good suggestion.  TLS only uses OSSL_PARAM_PUB_KEY.  This has to be in the format specified by the drafts (raw concatenation of public keys).  The import/export has multiple params, so isn't really the same.

paulidale replied:
That's a good suggestion and I've implemented it.
The format different is reasonable given that different param names are being used.

paulidale commented:
Updated in light of @t8m's comment.
Only the last commit is relevant.

paulidale commented:
Also converted an error message into two as per @beldmit's comment.

t8m commented:
I think the change is not negligible. IMO @slontis @beldmit should re-approve.

slontis reviewed (approved):
Looks like fairly small changes to me.
reapproved.

paulidale commented:
With the ML-KEM refactor merged, I've rebased and resolved the conflicts.  There were a duplication of some refactoring I did in `evp_test` which was noted by @vdukhovni ([here](https://github.com/openssl/openssl/pull/25990#discussion_r1882945213)).

vdukhovni commented:
My best guess is that the `IKME` secret will need to be split (which makes it possible to repeat it across multiple splits if needed).
For example, the ML-KEM provider currently (perhaps in error, but perhaps also safer) insists the the seed length be exactly the right size, it does not know what asking for a larger block of encap entropy might mean.

If some other algorithm also supports explicit encapsulation entropy, but it is a different size, then we can't combine them (in tests I guess, since non-test cases don't use the entropy).

Things are perhaps more interesting in keygen, where the `OSSL_PKEY_PARAM_ML_KEM_SEED` and `OSSL_PKEY_PARAM_EC_SEED` are both the same underlying string: `"seed"`. What happens when these need to be explicit, but different?

If we some day need to support the *(d, z)* private key format for `ML-KEM`, we may need to avoid overloading the keygen_seed interface to do so, but rather recognise this as an alternative private key format, under a different name, though ultimately still generate a key via the same low-level lib crypto functions.

Anyway I think we'll have some practical conflicts to resolve.

paulidale commented:
Yeah, there are going to be upcoming issues with respects to IKME and seeding.  We can either split them like we do for keys/secrets or we can reused the input key material (I did the latter) and check that the material is no shorter than what's required (& only using the amount required).

I'm deferring this particular decision to the PR that added a "proper" hybrid algorithm.

I expect there to be some other teething issues which will also need resolution at that time.

t-j-h commented:
You need to be able to query the length requirements for each algorithm and the hybrid needs to split things appropriately. The concept of hybrid is it should be able to be driven as logically one thing. So how to split needs to have the necessary information available within the hybrid implementation. Which means the underlying algorithms need to provide it. And the order must be determistic too for the hybrid.

paulidale commented:
@tjh, that's what's being done here.

t-j-h commented:
Where do you get the length of the ikme from for each algorithm?

paulidale commented:
It will go into the info[] if & when it's required -- IKME exists solely to support the ACVP tests and, given NIST's comments and stance, hybrid algorithms don't look like they are going to be able to be FIPS validated.  Therefore there is no need to support it at the moment.

vdukhovni commented:
> It will go into the info[] if & when it's required -- IKME exists solely to support the ACVP tests and, given NIST's comments and stance, hybrid algorithms don't look like they are going to be able to be FIPS validated. Therefore there is no need to support it at the moment.

I expect we'll still want to do KAT tests for the hybrids, the data might just not come from NIST>

vdukhovni commented:
Tried to build an X25519 + ML-KEM-768 hybrid.

* Had to lie to the TLS layer, to tell it that the provider "group" associated with this TLS group is "X25519", so the the right name would be be passed to both providers (ignored by ML-KEM, but needed in ECX)
* With that out of the way, a new error:

```
providers/implementations/keymgmt/hybrid_kmgmt.c:64: OpenSSL internal error: refcount error
```

baentsch commented:
fwiw, I completely agree with

> I'm deferring this particular decision to the PR that added a "proper" hybrid algorithm.

--> This PR is about resolving https://github.com/openssl/openssl/issues/25884 (generic hybrid capabilities) and I think this PR implements this fully (Thanks @paulidale !)

> I expect there to be some other teething issues which will also need resolution at that time.

Me too; so, in the interest of moving forward, can I ask whether there's still anything needed to get it over the finishing line (merge to feature branch) to then seriously work on the instantiation using ML-KEM (to eventually close out https://github.com/openssl/openssl/issues/17774)? 2 approvals are already there -- what about the change requests by @t8m @mattcaswell ? OK to mark resolved?

ACVP testing imo is even further out (e.g., part of the discussion in https://github.com/openssl/openssl/issues/26036 ?).

As a reminder: To the best of my knowledge, all of this is based on/for an [IETF draft](https://www.ietf.org/archive/id/draft-kwiatkowski-tls-ecdhe-mlkem-02.html) that's not finalized, albeit implemented by other code bases, so I think our focus for now should be to ascertain this works (interoperates). Full standards compliance can and should come as and when the standard is finally there (probably/possibly/hopefully together with KATs and other tests -- but I have no clue how IETF operates/what it demands of a proper, finalized RFC: If anyone reading knows, please chime in, ideally adding test requirements in a separate new issue (or https://github.com/openssl/openssl/issues/17774 which I suggest as a pre-existing issue that can be used for tracking this: @vavroch2010 , OK to add that to the backlog?).

baentsch commented:
> Tried to build an X25519 + ML-KEM-768 hybrid.

Thanks @vdukhovni ! Any chance you could share what you did, e.g. via draft PR so we all can take a look and move forward discussing (in that separate PR such as to close this one out)?

vdukhovni commented:
I just used the generic skeleton, pasting together the hybrid:
```
 include/internal/tlsgroups.h                                |  1 +
 providers/common/capabilities.c                             |  6 ++++++
 providers/defltprov.c                                       |  7 +++++++
 providers/implementations/include/prov/implementations.h    |  4 ++++
 providers/implementations/kem/build.info                    |  5 ++++-
 providers/implementations/kem/x25519_ml_kem_768_kem.c       | 30 ++++++++++++++++++++++++++++++
 providers/implementations/keymgmt/build.info                |  6 +++++-
 providers/implementations/keymgmt/x25519_ml_kem_768_kmgmt.c | 92 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 8 files changed, 149 insertions(+), 2 deletions(-)
```
This took very low code, mostly identical to the demo code for ECX + ECX, **but** I'm frankly sceptical that this *agnostic* approach which does not specifically model each hybrid's capabiilities, carefully mapping properties, ... is viable.
It may turn out that writing a few clean, more principled, *targetted* hybrids may prove more robust in terms of not getting confused about which properties go with which hybrid, how to combine those that apply to both, ...

t8m commented:
@paulidale please rebase this

mattcaswell commented:
See some fixes in https://github.com/openssl/openssl/pull/26220#issuecomment-2556761062 (the ref count patch there was courtesy of @vdukhovni)

mattcaswell commented:
There are problems with group name handling which needs a resolution. If you pass a group name down to the hybrid it should somehow split this up and pass the correct group name down to the constituent parts. At the moment it just passes down the whole group name unmodified which is not what you want.

vdukhovni commented:
Also, I'm at a complete loss as to why the provider-side key should bother with refcounts. There just aren't shared references to the thing at this layer.

t8m commented:
> There are problems with group name handling which needs a resolution. If you pass a group name down to the hybrid it should somehow split this up and pass the correct group name down to the constituent parts. At the moment it just passes down the whole group name unmodified which is not what you want.

I would not bother splitting this up. You just need to ignore the group as set from the TLS layer and set the group name param explicitly (or not set it at all) when generating params on the component side.

vdukhovni commented:
> I would not bother splitting this up. You just need to ignore the group as set from the TLS layer and set the group name param explicitly (or not set it at all) when generating params on the component side.


Yes!  This makes a lot more sense. The per-algorithm table in the provider is where the group names need to live, and when a parakeet with a group is being forwarded the group name should be replaced with the group for the algorithm, or entirely dropped, since the hybrid "group" is fiction.

vdukhovni reviewed (changes requested):
Significant structural issues remain, both memory management (valgrind reports issues) and correctness (handling GROUP parameters correctly per instance) as well as removing refcounts unless there's a clear compelling reason to have them. And more tests, including a working TLS test of at least one hybrid, before this should be merged.

vdukhovni commented:
> Significant structural issues remain, both memory management (valgrind reports issues) and correctness (handling GROUP parameters correctly per instance) as well as removing refcounts unless there's a clear compelling reason to have them. And more tests, including a working TLS test of at least one hybrid, before this should be merged.

Matt has addressed the memory issues (I took care of the immediate problem with the refcount, but I still think that refcounts should simply go away). So perhaps we're much closer but:

* I really want to see a test of ECX+ECX (X25519 wit X448)
    -  Using a suitable private TLS group code point that would not be present in the code seen  by users
        * do we have a way of doing runtime customisation of the default provider to optionally support test codepoints?
    - That overcomes the the "group name" parameter problem.
        * There is not a sensible combined group name parameter for a hybrid KEM that makes sense to pass to the underlying algorithms
        * Instead:
            * Each of the algorithm "info" records should have an optional (nullable) group name that replaces any group name parameter from the EVP layer with that value for just the underlying algorithm in question.
            * Deleting the parameter when the algorithm's group name is NULL.

paulidale commented:
TLS integration is out of scope.  I never intended to do that here.

I'm not sure what the problem is with the seeding.  The IKME stuff won't work, but as mentioned that's irrelevant since these won't be put through ACVP testing.  Key generation gets its input entropy directly from the RNG.  What's the problem?

The ref count stuff is copied from ECX code.  If it's not needed here, it's also not needed there.

vdukhovni commented:
> TLS integration is out of scope. I never intended to do that here.
> 
> I'm not sure what the problem is with the seeding. The IKME stuff won't work, but as mentioned that's irrelevant since these won't be put through ACVP testing. Key generation gets its input entropy directly from the RNG. What's the problem?
> 
> The ref count stuff is copied from ECX code. If it's not needed here, it's also not needed there.

* TLS integration was and is needed as a robust test of the correctness of the construct.
* Seeding is not JUST for ACVP tests, other hybrid constructs in the pipeline will have KAT tests.
* Also modifying params en-route to the underlying providers is needed not only for splitting the seed, but also for forwarding the correct group name or operating mode to each provider from the info tables.
* The ref count is in ECX because ECX keys are *legacy* keys, while the hybrid is purely provided, and so does not need refcounts.

paulidale commented:
Please include a question when flagging an OTC hold.

paulidale commented:
Rebased to fix conflicts.  Removed reference counting.

The ECX + ECX algorithm isn't permanent and is to be removed on merge.  All relevant commits are flagged `[TEMP]` for this reason.  There is consequently no need to add a TLS test for it.

I'll look into better keygen seeding.  Note that the referenced link reverses the order of the seed material compared to the encapsulation which is something not supported here.  This might require adding _combiner_ support earlier than anticipated.

vdukhovni commented:
> Note that the referenced link reverses the order of the seed material compared to the encapsulation which is something not supported here. This might require adding _combiner_ support earlier than anticipated.

Unclear what you have in mind. The IETF *hybrid design* I-D states that the wire-order of all the parts is the same as order of components in the **definition** of the algorithm (though not necessarily its **name**, IMNSHO, so I posted some strong comments about that to the IETF TLS group and the PR proposing renaming X25519MLKEM768 to reverse the name).

I would expect the order of seed components to match the wire order, regardless of what the algorithm is called. If you don't allow naming overrides so that something doing "X+Y" on the wire can't be called "YX" or "red_blue"... (i.e. naming bike shedding is independent of the technical details), then that's what should be changed. The seed order **SHOULD** match the wire order of the components.

paulidale commented:
> * Seeding is not JUST for ACVP tests, other hybrid constructs in the pipeline will have KAT tests.

What format is NIST going to use in their ACVP tests?

The IKME parameter was introduced for ACVP testing.  It should not be prematurely used for other KATs.  Until we know the format NIST is going to mandate, we should leave our options open.  This might mean another different parameter is required for non-NIST KATs.  At least when they are introduced.

vdukhovni commented:
> > * Seeding is not JUST for ACVP tests, other hybrid constructs in the pipeline will have KAT tests.
> 
> What format is NIST going to use in their ACVP tests?
> 
> The IKME parameter was introduced for ACVP testing. It should not be prematurely used for other KATs. Until we know the format NIST is going to mandate, we should leave our options open. This might mean another different parameter is required for non-NIST KATs. At least when they are introduced.

I don't expect NIST to be testing **generic** hybrid constructions. They're going to test  specific ones, and in tests for a specific construction, the pre-processing of the KAT test input can shuffle the entropy bits as needed, if a when needed.

The tests should not be the determining factor in building consistently implemented constructs in which (moniker aside) all the parts are in a consistent order.

t8m commented:
**Question for OTC:** Do we want a generic hybrid key management support?

t8m commented:
**OTC: Hybrid KEM for TLS does not require this.**
